src/main/java/org/opensearch/security/filter/SecurityFilter.java

```diff
@@ -384,6 +399,45 @@ private <Request extends ActionRequest, Response extends ActionResponse> void ap
             }
 
             PrivilegesEvaluationContext context = eval.createContext(user, action, request, task, injectedRoles);
+            User finalUser = user;
+            Consumer<PrivilegesEvaluatorResponse> handleUnauthorized = response -> {
+                auditLog.logMissingPrivileges(action, request, task);
+                String err;
+                if (!response.getMissingSecurityRoles().isEmpty()) {
+                    err = String.format("No mapping for %s on roles %s", finalUser, response.getMissingSecurityRoles());
+                } else {
+                    err = (injectedRoles != null)
+                        ? String.format(
+                            "no permissions for %s and associated roles %s",
+                            response.getMissingPrivileges(),
+                            context.getMappedRoles()
+                        )
+                        : String.format("no permissions for %s and %s", response.getMissingPrivileges(), finalUser);
+                }
+                log.debug(err);
+                listener.onFailure(new OpenSearchSecurityException(err, RestStatus.FORBIDDEN));
+            };
+
+            // NOTE: Since resource-access evaluation requires fetching documents from index, we make the call async otherwise it would
+            // require blocking transport threads leading to thread exhaustion and request timeouts
+            // We perform the rest of the evaluation as normal if the request is not for resource-access or if the feature is disabled
+            if (resourceAccessEvaluator.shouldEvaluate(request)) {
+                resourceAccessEvaluator.evaluateAsync(request, action, context, ActionListener.wrap(response -> {
+                    if (handlePermissionCheckRequest(listener, response, action)) {
```

Comment:
its not clear from looking at this that the response will be sent from within this method if it is a dry run.